### PR TITLE
[compiler] Type inference for tagged template literals

### PR DESCRIPTION
**Pull request by josephsavona**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #30868
* __->__ #30869

At Meta we have a pattern of using tagged template literals for features that are compiled away:

```
// Relay:
graphql`...graphql text...`
```

In many cases these tags produce a primitive value, and we can get even more optimal output if we can tell the compiler about these types. The new moduleTypeProvider gives us the ability to declare such types, this PR extends the compiler to use this type information for TaggedTemplateExpression values.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructuring-mixed-scope-and-local-variables-with-default.expect.md [hunk 1]
```diff
@@ -63,67 +63,63 @@ function useFragment(_arg1, _arg2) {
 }
 
 function Component(props) {
-  const $ = _c(9);
-  let t0;
-  if ($[0] === Symbol.for("react.memo_cache_sentinel")) {
-    t0 = graphql`
+  const $ = _c(8);
+  const post = useFragment(
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/__tests__/fixtures/compiler/destructuring-mixed-scope-and-local-variables-with-default.expect.md:67]: this fixture has a locally defined `useFragment` so it isn't using the type definition for the useFragment in shared-runtime.

File: compiler/packages/babel-plugin-react-compiler/src/TypeInference/InferTypes.ts [hunk 2]
```diff
@@ -250,6 +250,7 @@ function* generateInstructionTypes(
     }
 
     case 'CallExpression': {
+      const returnType = makeType();
```
Comment (mofeiZ) [inline compiler/packages/babel-plugin-react-compiler/src/TypeInference/InferTypes.ts:253]: Hmm I don't quite understand -- what does creating an intermediate `returnType` do here?

**Review by mofeiZ — APPROVED** [review]
Nice!

File: compiler/packages/babel-plugin-react-compiler/src/TypeInference/InferTypes.ts (hunk 2 above)
Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/TypeInference/InferTypes.ts:253]: It doesn't really matter in practice, but this ensures we have a place to put the inferred return type of the function (eg from its signature) separately from what we infer about how it's lvalue is used. For example in `foo() && ...`, we would inver the lvalue of the `foo()` call is a primitive bc of the logical expression, but that means the actual call expression's return type information is lost. This way we can still track the actual function return type.

**Pull request merged**